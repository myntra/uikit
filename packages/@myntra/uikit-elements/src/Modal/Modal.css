#--uikit-modal-container-- {
  position: fixed;
  top: 0;
  left: 0;
  width: 0;
  height: 0;
  z-index: var(--z-index-modal, 2000);
}

#--uikit-modal-container--:empty {
  display: none;
}
