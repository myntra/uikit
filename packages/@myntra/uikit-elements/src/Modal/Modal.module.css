.modal {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: var(--size-large);
  z-index: var(--z-index-modal);
}

.backdrop {
  background: rgba(0, 0, 0, 0.54);
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: -1;
}

.body {
  position: relative;
}

.content {
  max-height: 80vh;
  overflow: auto;
}

.wrapper {
  padding: var(--size-large);
  min-width: 300px;
  background: white;
}

.actions {
  padding-top: var(--size-large);
  margin-top: calc(2 * var(--size-large));
  border-top: 1px solid var(--color-gray-dark);
  text-align: right;
}

.close {
  position: absolute;
  top: -13px;
  right: -25px;
  padding-right: 0;
  color: var(--typography-color-light-primary);
}
