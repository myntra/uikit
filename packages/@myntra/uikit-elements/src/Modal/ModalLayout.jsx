import React from 'react'
import PropTypes from 'prop-types'
import { classnames } from '@myntra/uikit-utils'
import styles from './Modal.module.css'

/**
 @since 0.3.0
 @status EXPERIMENTAL
 @example
 <Modal.Layout title="Example Modal" actions={close => <Button onClick={close}>Okay</Button>}>
   <p>A simple modal with a form.</p>

   <Form.Text label="Name" description="Your name."/>

   <p>You can render anything here!!</p>

   <Image src="//picsum.photos/200" height={200}  width={200} />
 </Modal.Layout>
 */
function ModalLayout({ title, actions, children, close }) {
  return (
    <div className={classnames('wrapper').use(styles)}>
      {title && <h1>{title}</h1>}

      {children}

      {actions && (
        <div className={classnames('actions').use(styles)}>
          {typeof actions === 'function' ? actions(close) : actions}
        </div>
      )}
    </div>
  )
}

ModalLayout.propTypes = {
  /** Contents */
  children: PropTypes.any.isRequired,
  /** Modal title */
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  /** Action buttons */
  actions: PropTypes.oneOfType([PropTypes.func, PropTypes.element, PropTypes.arrayOf(PropTypes.element)]),
  /** Trigger close action */
  close: PropTypes.func
}

export default ModalLayout
