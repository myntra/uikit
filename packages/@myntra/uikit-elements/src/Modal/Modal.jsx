import React, { PureComponent, Fragment } from 'react'
import PropTypes from 'prop-types'
import { Button, Portal } from '../index.js'

import ModalLayout from './ModalLayout'
import classnames from './Modal.module.css'

// eslint-disable-next-line
const FragmentWithFallback = Fragment || (({ children }) => <div style={{ display: 'contents' }}>{children}</div>)

/**
 A component to display popup modal.

 @since 0.3.0
 @status EXPERIMENTAL
 @example
 <Modal trigger="Open Modal" title="Example Modal" actions={close => <Button onClick={close}>Okay</Button>}>
   <p>A simple modal with a form.</p>

   <Form.Text label="Name" description="Your name."/>

   <p>You can render anything here!!</p>

   <Image src="//picsum.photos/200" height={200}  width={200} />
 </Modal>
 */
class Modal extends PureComponent {
  static propTypes = {
    /** Contents of the popup modal */
    children: PropTypes.any.isRequired,
    /** @private */
    className: PropTypes.string,
    /** Trigger to open modal */
    trigger: PropTypes.oneOfType([PropTypes.string.isRequired, PropTypes.element.isRequired]),
    /** Modal title */
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    /** Action buttons */
    actions: PropTypes.oneOfType([PropTypes.func, PropTypes.element, PropTypes.arrayOf(PropTypes.element)]),
    /** Dropdown state */
    isOpen: PropTypes.bool,
    /** To hide close button */
    hideClose: PropTypes.bool,
    /**
     * Event fired when dropdown drawer is displayed
     * @function
     */
    onOpen: PropTypes.func,
    /**
     * Event fired when dropdown drawer is closed
     * @function
     */
    onClose: PropTypes.func,
    /**
     * Render a custom wrapper.
     */
    render: PropTypes.func
  }

  static defaultProps = {
    render: ModalLayout
  }

  state = {
    isOpen: false
  }

  handleOpen = () => {
    this.setState({ isOpen: true })
    this.props.onOpen && this.props.onOpen(true)
  }

  handleClose = () => {
    this.setState({ isOpen: false })
    this.props.onClose && this.props.onClose(false)
  }

  render() {
    const { trigger, children, isOpen, render, title, actions, hideClose } = this.props

    const content = (
      <div className={classnames('modal')}>
        <div className={classnames('backdrop')} onClick={this.handleClose} />
        <div className={classnames('body')}>
          <div className={classnames('content')}>{render({ title, actions, children, close: this.handleClose })}</div>

          {hideClose ? null : (
            <Button className={classnames('close')} type="link" icon="times" title="close" onClick={this.handleClose} />
          )}
        </div>
      </div>
    )

    return (
      <FragmentWithFallback>
        {typeof trigger === 'string' ? (
          <Button label={trigger} secondaryIcon="chevron-down" onClick={this.handleOpen} />
        ) : trigger ? (
          React.cloneElement(trigger, { onClick: this.handleOpen })
        ) : null}
        {(typeof isOpen === 'boolean' ? isOpen : this.state.isOpen) && <Portal container>{content}</Portal>}
      </FragmentWithFallback>
    )
  }
}

Modal.Layout = ModalLayout

export default Modal
