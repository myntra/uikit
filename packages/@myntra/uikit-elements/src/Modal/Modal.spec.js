import React from 'react'
import { shallow, mount } from 'enzyme'
import Modal from './Modal'

describe('Modal', () => {
  it('renders', () => {
    const wrapper = shallow(
      <Modal isOpen trigger={<button>Trigger</button>} title="Title" actions={<button>Action</button>}>
        <span>Some</span>
        <p>Text</p>
      </Modal>
    )

    expect(wrapper).toMatchSnapshot()
  })
  it('fires open/close events', () => {
    const onOpen = jest.fn()
    const onClose = jest.fn()
    const wrapper = mount(
      <Modal trigger="Trigger" onOpen={onOpen} onClose={onClose}>
        <span>Some</span>
        <p>Text</p>
      </Modal>
    )

    expect(wrapper.text()).toBe('Trigger')
    expect(wrapper.find('.modal')).toHaveLength(0)
    expect(onOpen).toHaveBeenCalledTimes(0)
    expect(onClose).toHaveBeenCalledTimes(0)

    wrapper.find('Button').simulate('click')
    expect(onOpen).toHaveBeenCalledTimes(1)
    expect(onClose).toHaveBeenCalledTimes(0)
    expect(wrapper.find('.modal')).toHaveLength(1)
    expect(wrapper.find('.modal').text()).toBe('SomeText')

    wrapper.find('.backdrop').simulate('click')
    expect(onOpen).toHaveBeenCalledTimes(1)
    expect(onClose).toHaveBeenCalledTimes(1)
    expect(wrapper.find('.modal')).toHaveLength(0)
  })

  it('should not render close button for hideClose prop', () => {
    const wrapper = mount(
      <Modal trigger="Trigger" hideClose>
        <span>Some Text</span>
      </Modal>
    )
    expect(wrapper.find('.close')).toHaveLength(0)
  })
})
