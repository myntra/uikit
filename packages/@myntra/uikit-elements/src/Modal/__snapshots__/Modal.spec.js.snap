// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders 1`] = `
<Dropdown
  approxContentHeight={320}
  approxContentWidth={240}
  container={
    <div
      id="--uikit-modal-container--"
    />
  }
  isOpen={true}
  onOpen={[Function]}
  trigger={
    <button>
      Trigger
    </button>
  }
  useClickAway={false}
>
  <div
    className="modal"
  >
    <div
      className="backdrop"
      onClick={[Function]}
    />
    <div
      className="body"
    >
      <div
        className="content"
      >
        <div
          className="wrapper"
        >
          <h1>
            Title
          </h1>
          <span>
            Some
          </span>
          <p>
            Text
          </p>
          <div
            className="actions"
          >
            <button>
              Action
            </button>
          </div>
        </div>
      </div>
      <Button
        className="close"
        disabled={false}
        icon="times"
        onClick={[Function]}
        title="close"
        type="link"
      />
    </div>
  </div>
</Dropdown>
`;
